release/releaser.go

```diff
@@ -45,539 +36,318 @@ func NewReleaser(
-		releaseType = "release_one_without_update"
-		actions, err = r.releaseWithoutUpdate(releaseType, msg, inst, services)
-
-	default:
-		releaseType = "release_one"
-		actions, err = r.releaseImages(releaseType, msg, inst, services, images)
+	// Preparation: we always need the repository
+	rc := NewReleaseContext(inst)
+	defer rc.Clean()
+	logStatus("Cloning git repository.")
+	timer = r.metrics.NewStageTimer("clone_repository")
+	if err = rc.CloneRepo(); err != nil {
+		return nil, err
 	}
-	return releaseType, actions, err
-}
-
-func (r *Releaser) releaseImages(method, msg string, inst *instance.Instance, getServices ServiceSelector, getImages ImageSelector) ([]ReleaseAction, error) {
-	var res []ReleaseAction
-	res = append(res, r.releaseActionPrintf(msg))
-
-	var (
-		base  = r.metrics.StageDuration.With("method", method)
-		stage *metrics.Timer
-	)
+	timer.ObserveDuration()
 
-	defer func() { stage.ObserveDuration() }()
-	stage = metrics.NewTimer(base.With("stage", "fetch_platform_services"))
+	// From here in, we collect the results of the calculations.
+	results := flux.ReleaseResult{}
 
-	services, err := getServices.SelectServices(inst)
+	// Figure out the services involved.
+	logStatus("Finding defined services.")
+	timer = r.metrics.NewStageTimer("select_services")
+	var updates []*ServiceUpdate
+	updates, err = selectServices(rc, &spec, results, logStatus)
+	timer.ObserveDuration()
 	if err != nil {
-		return nil, errors.Wrap(err, "fetching platform services")
-	}
-	if len(services) == 0 {
-		res = append(res, r.releaseActionPrintf("No selected services found. Nothing to do."))
-		return res, nil
+		return nil, err
 	}
-
-	stage.ObserveDuration()
-	stage = metrics.NewTimer(base.With("stage", "calculate_applies"))
-
-	// Each service is running multiple images.
-	// Each image may need to be upgraded, and trigger an apply.
-	images, err := getImages.SelectImages(inst, services)
-	if err != nil {
-		return nil, errors.Wrap(err, "collecting available images to calculate applies")
+	logStatus("Found %d services.", len(updates))
+	report(results)
+
+	logStatus("Looking up images.")
+	if spec.ImageSpec != flux.ImageSpecNone {
+		timer = r.metrics.NewStageTimer("lookup_images")
+		// Figure out how the services are to be updated.
+		updates, err = calculateImageUpdates(rc.Instance, updates, &spec, results, logStatus)
+		timer.ObserveDuration()
+		if err != nil {
+			return nil, err
+		}
+		report(results)
 	}
 
-	updateMap := CalculateUpdates(services, images, func(format string, args ...interface{}) {
-		res = append(res, r.releaseActionPrintf(format, args...))
-	})
-
-	if len(updateMap) <= 0 {
-		res = append(res, r.releaseActionPrintf("All selected services are running the requested images. Nothing to do."))
-		return res, nil
+	// At this point we have have filtered the updates we can do down
+	// to nothing. Check and exit early if so
+	if len(updates) == 0 {
+		logStatus("No updates to do, finishing.")
```

Comment:
If it's a "no update change", we will not have executed the previous block which looks for new images and filters `updates` down to just those things with image updates.

Is the layout of this code a bit misleading, perhaps -- I can try and rejig it.